hclsyntax/expression.go

```diff
@@ -631,7 +632,7 @@ func (e *IndexExpr) Value(ctx *hcl.EvalContext) (cty.Value, hcl.Diagnostics) {
 	diags = append(diags, collDiags...)
 	diags = append(diags, keyDiags...)
 
-	val, indexDiags := hcl.Index(coll, key, &e.SrcRange)
+	val, indexDiags := hcl.Index(coll, key, &e.BracketRange)
```

Comment:
This final argument is what `hcl.Index` uses as the subject of any diagnostics it produces if the index operation fails. I've changed this so that errors relating to incorrectly-typed or incorrectly-valued indices will continue to highlight only the bracketed part, making it clearer which part of the expression is faulty.